src/NumericalAlgorithms/Spectral/SwshDerivatives.hpp

```diff
@@ -305,5 +305,168 @@ using SwshDerivatives = detail::SwshDerivativesImpl<
+ */
+template <int Spin>
+auto swsh_buffer(const size_t l_max,
+                 const size_t number_of_radial_points) noexcept {
+  return SpinWeighted<ComplexModalVector, Spin>{
+      size_of_libsharp_coefficient_vector(l_max) * number_of_radial_points};
+}
+
+namespace detail {
+// template 'implementation' for the `swsh_derivatives` function below which
+// evaluates an arbitrary number of derivatives, and places them in the set of
+// nodal containers passed by pointer.
+template <ComplexRepresentation Representation, typename... DerivativeKinds,
+          typename... ArgumentTypes, size_t... Is>
+void swsh_derivatives_impl(
+    const size_t l_max, const size_t number_of_radial_points,
+    std::index_sequence<Is...> /*meta*/,
+    tmpl::list<DerivativeKinds...> /*meta*/,
+    cpp17::bool_constant<true> /*buffers_included_in_arguments*/,
+    const std::tuple<ArgumentTypes...>& argument_tuple) noexcept {
+  SwshDerivatives<
+      tmpl::list<Tags::Derivative<
+          ::Tags::SpinWeighted<
+              ::Tags::TempScalar<Is, ComplexDataVector>,
+              std::integral_constant<
+                  int, std::decay_t<decltype(get<Is + 3 * sizeof...(Is)>(
+                           argument_tuple))>::spin>>,
+          DerivativeKinds>...>,
+      Representation>::apply_to_vectors(get<Is>(argument_tuple)...,
+                                        get<Is + sizeof...(Is)>(
+                                            argument_tuple)...,
+                                        get<Is + 2 * sizeof...(Is)>(
+                                            argument_tuple)...,
+                                        get<Is + 3 * sizeof...(Is)>(
+                                            argument_tuple)...,
+                                        l_max, number_of_radial_points);
+}
+
+template <ComplexRepresentation Representation, typename... DerivativeKinds,
+          typename... ArgumentTypes, size_t... Is>
+void swsh_derivatives_impl(
+    const size_t l_max, const size_t number_of_radial_points,
+    std::index_sequence<Is...> index_sequence,
+    tmpl::list<DerivativeKinds...> derivative_kinds,
+    cpp17::bool_constant<false> /*buffers_included_in_arguments*/,
+    const std::tuple<ArgumentTypes...>& argument_tuple) noexcept {
+  auto derivative_buffer_tuple = std::make_tuple(
+      swsh_buffer<std::decay_t<decltype(*get<Is>(argument_tuple))>::spin>(
+          l_max, number_of_radial_points)...);
+  auto input_buffer_tuple = std::make_tuple(
+      swsh_buffer<std::decay_t<decltype(get<Is + sizeof...(Is)>(
+          argument_tuple))>::spin>(l_max, number_of_radial_points)...);
+  swsh_derivatives_impl<Representation>(
+      l_max, number_of_radial_points, index_sequence, derivative_kinds,
+      cpp17::bool_constant<true>{},
+      std::make_tuple(make_not_null(&get<Is>(derivative_buffer_tuple))...,
+                      make_not_null(&get<Is>(input_buffer_tuple))...,
+                      get<Is>(argument_tuple)...,
```

Comment:
uh oh, lots of copying here. Seems expensive

Reply:
ah, trouble. I had thought `std::make_tuple` would send along the references, but it looks like it decays the type. Would I be correct in thinking `forward_as_tuple` will do what I'm looking for?

Reply:
I think that'll do what you want. I would run the result through `TypeDisplayer` to get the compiler to spit out an error with the type, but I believe that should do what you want.